File: internal/domain/constraints.go
package domain

import "fmt"

const (
	MaxStickerSide            = 512
	MaxStickerFPS             = 30
	MaxStickerDurationSeconds = 3
	MaxStickerSizeBytes       = 256 * 1024
	DefaultImageFPS           = 30
	DefaultImageDuration      = 3
	StaticStickerSide         = 512
	EmojiSide                 = 100
)

type Size struct {
	Width  int
	Height int
}

func ScaleToFit(src Size, maxSide int) (Size, error) {
	if src.Width <= 0 || src.Height <= 0 {
		return Size{}, fmt.Errorf("invalid size: %dx%d", src.Width, src.Height)
	}

	if src.Width == src.Height {
		return Size{Width: maxSide, Height: maxSide}, nil
	}

	if src.Width > src.Height {
		height := int(float64(src.Height) * float64(maxSide) / float64(src.Width))
		if height <= 0 {
			height = 1
		}
		return Size{Width: maxSide, Height: height}, nil
	}

	width := int(float64(src.Width) * float64(maxSide) / float64(src.Height))
	if width <= 0 {
		width = 1
	}
	return Size{Width: width, Height: maxSide}, nil
}

File: python/core/constraints.py
from __future__ import annotations

from dataclasses import dataclass

MAX_STICKER_SIDE = 512
MAX_STICKER_FPS = 30
MAX_STICKER_DURATION_SECONDS = 3
MAX_STICKER_SIZE_BYTES = 256 * 1024
DEFAULT_IMAGE_FPS = 30
DEFAULT_IMAGE_DURATION = 3
STATIC_STICKER_SIDE = 512
EMOJI_SIDE = 100


@dataclass(frozen=True)
class Size:
    width: int
    height: int


def scale_to_fit(src: Size, max_side: int) -> Size:
    if src.width <= 0 or src.height <= 0:
        raise ValueError(f"invalid size: {src.width}x{src.height}")

    if src.width == src.height:
        return Size(width=max_side, height=max_side)

    if src.width > src.height:
        height = int(float(src.height) * float(max_side) / float(src.width))
        if height <= 0:
            height = 1
        return Size(width=max_side, height=height)

    width = int(float(src.width) * float(max_side) / float(src.height))
    if width <= 0:
        width = 1
    return Size(width=width, height=max_side)
